src/AbstractCollection.php

```diff
@@ -274,7 +274,15 @@ public function valid() {
 	 * @internal
 	 */
 	public function __clone() {
-		return new static($this->collection);
+		$clonedCollection = [];
+		foreach ($this->collection as $key => $element) {
+			if (is_object($element)) {
+				$clonedCollection[$key] = clone $element;
+				continue;
+			}
+			$clonedCollection[$key] = $element;
+		}
+		$this->collection = $clonedCollection;
```

Comment:
It should rather be `return new static($clonedCollection);` - right? Internal collections should be treated "immutable", like only mutator methods will mutate it but not `clone`

Reply:
If PHP was a "normal" language, it should be so :smile:

If I write `return new static($clonedCollection);` I obtain exactly what I don't want: a clone of the collection, containing references to the original collected objects, while I'ld like a clone of the collection, containg cloned objects.
According to [Php documentation](https://www.php.net/manual/en/language.oop5.cloning.php), the `__clone` method is called on the newly created object "*to allow any necessary properties that need to be changed.*" (which behavior you'ld like to avoid).